src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,958 @@
+                                  std::is_same<BsonT, bsoncxx::types::b_array>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_binary>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_oid>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_bool>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_date>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_int32>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_int64>::value ||
+                                  std::is_same<BsonT, bsoncxx::oid>::value;
+};
+
+/**
+ * A templated struct containing a bool value that specifies whether the
+ * provided template parameter is a BSON type that contains a view.
+ */
+template <class BsonT>
+struct is_bson_view {
+    static constexpr auto value = std::is_same<BsonT, bsoncxx::types::b_utf8>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_document>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_array>::value ||
+                                  std::is_same<BsonT, bsoncxx::types::b_binary>::value;
+};
+
+class BSONOutputArchive : public OutputArchive<BSONOutputArchive> {
+    /**
+    * The possible states for the BSON nodes being output by the archive.
+    */
+    enum class OutputNodeType { StartObject, InObject, StartArray, InArray };
+
+    typedef bsoncxx::builder::core BSONBuilder;
+
+   public:
+    /**
+    * Construct a BSONOutputArchive that will output to the provided stream.
+    *
+    * @param stream
+    *   The stream to which the archiver will output BSON data.
+    */
+    BSONOutputArchive(std::ostream& stream)
+        : OutputArchive<BSONOutputArchive>{this},
+          _bsonBuilder{false},
+          _writeStream{stream},
+          _nextName{nullptr} {
+    }
+
+   private:
+    /**
+     * Writes the current contents of the BSON document builder to the
+     * output stream.
+     */
+    void writeDoc() {
+        _writeStream.write(reinterpret_cast<const char*>(_bsonBuilder.view_document().data()),
+                           _bsonBuilder.view_document().length());
+        _bsonBuilder.clear();
+    }
+
+   public:
+    /**
+     * Checks if the most recent object written was in the root of the document.
+     * If so, that document is written to the archive.
+     */
+    void writeDocIfRoot() {
+        if (_nodeTypeStack.empty()) {
+            writeDoc();
+        }
+    }
+
+    /**
+     * Starts a new node in the BSON output.
+     *
+     * The node can optionally be given a name by calling setNextName prior to
+     * creating the node. Nodes only need to be started for types that are
+     * objects or arrays.
+     */
+    void startNode() {
+        writeName(true);
```

Comment:
Does this need to do a check here for root? I think always passing true works just because of how `writeName()` works, but it seems to me like this could pass in `true` even when it's not opening an element at the root.